x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryTranslator.java

```diff
@@ -535,104 +528,122 @@ protected LeafAgg toAgg(String id, Count c) {
+            return new MinAgg(id, asFieldOrLiteralOrScript(m));
         }
     }
 
     static class MADs extends SingleValueAggTranslator<MedianAbsoluteDeviation> {
         @Override
         protected LeafAgg toAgg(String id, MedianAbsoluteDeviation m) {
-            return new MedianAbsoluteDeviationAgg(id, field(m));
+            return new MedianAbsoluteDeviationAgg(id, asFieldOrLiteralOrScript(m));
         }
     }
 
     static class Firsts extends TopHitsAggTranslator<First> {
 
         @Override
         protected LeafAgg toAgg(String id, First f) {
-            return new TopHitsAgg(id, topAggsField(f, f.field()), f.dataType(),
-                topAggsField(f, f.orderField()), f.orderField() == null ? null : f.orderField().dataType(), SortOrder.ASC);
+            return new TopHitsAgg(
+                id,
+                asFieldOrLiteralOrScript(f, f.field()),
+                f.dataType(),
+                asFieldOrLiteralOrScript(f, f.orderField()),
+                f.orderField() == null ? null : f.orderField().dataType(),
+                SortOrder.ASC
+            );
         }
     }
 
     static class Lasts extends TopHitsAggTranslator<Last> {
 
         @Override
         protected LeafAgg toAgg(String id, Last l) {
-            return new TopHitsAgg(id, topAggsField(l, l.field()), l.dataType(),
-                topAggsField(l, l.orderField()), l.orderField() == null ? null : l.orderField().dataType(), SortOrder.DESC);
+            return new TopHitsAgg(
+                id,
+                asFieldOrLiteralOrScript(l, l.field()),
+                l.dataType(),
+                asFieldOrLiteralOrScript(l, l.orderField()),
+                l.orderField() == null ? null : l.orderField().dataType(),
+                SortOrder.DESC
+            );
         }
     }
 
     static class StatsAggs extends CompoundAggTranslator<Stats> {
 
         @Override
         protected LeafAgg toAgg(String id, Stats s) {
-            return new StatsAgg(id, field(s));
+            return new StatsAgg(id, asFieldOrLiteralOrScript(s));
         }
     }
 
     static class ExtendedStatsAggs extends CompoundAggTranslator<ExtendedStats> {
 
         @Override
         protected LeafAgg toAgg(String id, ExtendedStats e) {
-            return new ExtendedStatsAgg(id, field(e));
+            return new ExtendedStatsAgg(id, asFieldOrLiteralOrScript(e));
         }
     }
 
     static class MatrixStatsAggs extends CompoundAggTranslator<MatrixStats> {
 
         @Override
         protected LeafAgg toAgg(String id, MatrixStats m) {
-            return new MatrixStatsAgg(id, singletonList(field(m)));
+            if (isFieldOrLiteral(m.field())) {
+                return new MatrixStatsAgg(id, singletonList(field(m, m.field())));
+            }
+            throw new SqlIllegalArgumentException(
+                "Cannot use scalar functions or operators: [{}] in matrix stats " + "aggregate functions [KURTOSIS] and [SKEWNESS]",
```

Comment:
Can this error message be improved?
- why is it mentioning matrix stats, isn't this an implementation detail?
- not a biggie and can be left as is, but can the message be more targeted to the exact aggregate function? (mention only KURTOSIS or SKEWNESS)

Reply:
> * why is it mentioning matrix stats, isn't this an implementation detail?

Yep, I can remove that.

> * not a biggie and can be left as is, but can the message be more targeted to the exact aggregate function? (mention only KURTOSIS or SKEWNESS)

Not really, I've tried it,  because we don't have the original function anymore here, only `MatrixStats`. I was considering checking that earlier on the functions themselves, but then `resolveType()` didn't seem to me a good place to add such a verification, since it's regarding the DataType and not if the arg is field or function.

Reply:
This is better handled in the verifier by checking that `Kurtosis` & co (namely `MatrixStatsEnclosed`) don't have a scalar expression as argument.
Further more I would add an issue to extend the aggregation to support scripts not just fields.

Reply:
Created issue: #55344